packages/vira/src/elements/vira-dropdown.element.ts

```diff
@@ -0,0 +1,511 @@
+    },
+    cleanupCallback({state, updateState}) {
+        updateState({showPopUpResult: undefined});
+        state.popUpManager.destroy();
+    },
+    initCallback({state, updateState, host, inputs, dispatch, events}) {
+        state.popUpManager.listen(HidePopUpEvent, () => {
+            updateState({showPopUpResult: undefined});
+            if (!inputs.isDisabled) {
+                const dropdownWrapper = host.shadowRoot.querySelector('.dropdown-wrapper');
+
+                assertInstanceOf(
+                    dropdownWrapper,
+                    HTMLButtonElement,
+                    'failed to find dropdown wrapper child',
+                );
+
+                dropdownWrapper.focus();
+            }
+        });
+        state.popUpManager.listen(NavSelectEvent, (event) => {
+            const optionIndex = event.detail.x;
+            const option = inputs.options[optionIndex];
+            if (!option) {
+                throw new Error(`Found no dropdown option at index '${optionIndex}'`);
+            }
+            /** Only close upon option selection if the dropdown is not multi select. */
+            if (!inputs.isMultiSelect) {
+                triggerPopUpState(
+                    {emitEvent: true, open: false},
+                    {
+                        dispatch: (openState) => {
+                            dispatch(new events.openChange(openState));
+                        },
+                        host,
+                        popUpManager: state.popUpManager,
+                        updateState,
+                    },
+                );
+            }
+
+            dispatch(
+                new events.selectedChange(
+                    createNewSelection(option.id, inputs.selected, !!inputs.isMultiSelect),
+                ),
+            );
+        });
+        updateState({navController: new NavController(host)});
+    },
+    renderCallback({dispatch, events, state, inputs, updateState, host}) {
+        assertUniqueIdProps(inputs.options);
+
+        function triggerPopUp(param: Parameters<typeof triggerPopUpState>[0]) {
+            triggerPopUpState(param, {
+                dispatch: (openState) => {
+                    dispatch(new events.openChange(openState));
+                },
+                host,
+                popUpManager: state.popUpManager,
+                updateState,
+            });
+        }
+
+        if (inputs.isDisabled) {
+            triggerPopUp({open: false, emitEvent: false});
+        } else if (inputs.z_debug_forceOpenState != undefined) {
+            if (!inputs.z_debug_forceOpenState && state.showPopUpResult) {
+                triggerPopUp({emitEvent: false, open: false});
+            } else if (inputs.z_debug_forceOpenState && !state.showPopUpResult) {
+                triggerPopUp({emitEvent: false, open: true});
+            }
+        }
+
+        const selectedOptions: ReadonlyArray<Readonly<ViraDropdownOption>> = inputs.selected.length
```

Comment:
I think this should be moved inside a function and that function should have unit tests written for it.

Reply:
done